File: apps/command-center-v3/src/components/options/optionsLifecyclePrimary.ts
/** Same order as scripts/lib/options_lifecycle_primary.py. One primary bucket. */
export type Bucket = 'blocked' | 'action_now' | 'harvest' | 'defend' | 'expiry' | 'mature'

const ORDER: Bucket[] = ['blocked', 'action_now', 'harvest', 'defend', 'expiry', 'mature']

export function matchingBuckets(p: any): Bucket[] {
  const rec = String(p?.decision?.recommendation || '')
  const dteRaw = p?.economics?.dte_nearest
  const dte = dteRaw == null || Number.isNaN(Number(dteRaw)) ? 99 : Number(dteRaw)
  const found: Bucket[] = []
  if (rec === 'DATA_BLOCKED') found.push('blocked')
  if (p?.decision?.urgency === 'red') found.push('action_now')
  if (rec.startsWith('HARVEST')) found.push('harvest')
  if (rec === 'DEFEND' || rec === 'ROLL') found.push('defend')
  if (dte <= 7 || rec === 'ACCEPT_ASSIGNMENT' || rec === 'EXERCISE_REVIEW') found.push('expiry')
  if (rec === 'LET_MATURE' || rec === 'HOLD') found.push('mature')
  return found
}

export function primaryBucket(p: any): Bucket {
  const found = matchingBuckets(p)
  for (const key of ORDER) if (found.includes(key)) return key
  return 'mature'
}

export function subordinateBuckets(p: any): Bucket[] {
  const primary = primaryBucket(p)
  return matchingBuckets(p).filter(k => k !== primary)
}

File: scripts/lib/options_lifecycle_primary.py
"""One primary lifecycle bucket per strategy. Other matches are subordinate."""
from __future__ import annotations

from typing import Any

ORDER = ("blocked", "action_now", "harvest", "defend", "expiry", "mature")


def _matches(position: dict[str, Any]) -> list[str]:
    decision = position.get("decision") or {}
    rec = str(decision.get("recommendation") or "")
    econ = position.get("economics") or {}
    dte = econ.get("dte_nearest")
    try:
        dte_n = float(dte) if dte is not None else 99.0
    except (TypeError, ValueError):
        dte_n = 99.0
    found = []
    if rec == "DATA_BLOCKED":
        found.append("blocked")
    if decision.get("urgency") == "red":
        found.append("action_now")
    if rec.startswith("HARVEST"):
        found.append("harvest")
    if rec in {"DEFEND", "ROLL"}:
        found.append("defend")
    if dte_n <= 7 or rec in {"ACCEPT_ASSIGNMENT", "EXERCISE_REVIEW"}:
        found.append("expiry")
    if rec in {"LET_MATURE", "HOLD"}:
        found.append("mature")
    return found


def primary_bucket(position: dict[str, Any]) -> str:
    found = _matches(position)
    for key in ORDER:
        if key in found:
            return key
    return "mature"


def subordinate_buckets(position: dict[str, Any]) -> list[str]:
    primary = primary_bucket(position)
    return [k for k in _matches(position) if k != primary]
